chain/network/src/peer_manager/peer_manager_actor.rs

```diff
@@ -2127,14 +2133,92 @@ impl PeerManagerActor {
-        *self.state.chain_info.write() = info.0;
+        let now = self.clock.now_utc();
+        let info = info.0;
+        let state = self.state.clone();
+        let event_sink = self.event_sink.clone();
+        tokio::spawn(async move {
+            *state.chain_info.write() = info.clone();
+            // If the key set didn't change, early exit.
+            if !state.accounts_data.set_keys(info.tier1_accounts.clone()) {
+                return;
+            }
+            // If the set of keys has changed, and the node is a validator,
+            // we should try to sign data and broadcast it. However, this is
+            // also a trigger for a full sync, so a dedicated broadcast is
+            // not required.
+            //
+            // TODO(gprusak): For dynamic self-IP-discovery, add a STUN daemon which
+            // will add new AccountData and trigger an incremental broadcast.
+            if let Some(vc) = &state.config.validator {
+                let my_account_id = vc.signer.validator_id();
+                let my_public_key = vc.signer.public_key();
+                // TODO(gprusak): STUN servers should be queried periocally by a daemon
+                // so that the my_peers list is always resolved.
+                // Note that currently we will broadcast an empty list.
+                // It won't help us to connect the the validator BUT it
+                // will indicate that a validator is misconfigured, which
+                // is could be useful for debugging. Consider keeping this
+                // behavior for situations when the IPs are not known.
+                let my_peers = match &vc.endpoints {
+                    config::ValidatorEndpoints::TrustedStunServers(_) => vec![],
+                    config::ValidatorEndpoints::PublicAddrs(addrs) => {
+                        addrs.iter().cloned().map(|addr| PeerAddr { addr, peer_id: None }).collect()
+                    }
+                };
+                let my_data = info.tier1_accounts.iter().filter_map(|((epoch_id,account_id),key)| {
+                    if account_id != my_account_id{
+                        return None;
+                    }
+                    if key != &my_public_key {
+                        warn!(target: "network", "node's account_id found in TIER1 accounts, but the public keys do not match");
+                        return None;
+                    }
+                    Some(AccountData {
+                        epoch_id: epoch_id.clone(),
+                        account_id: my_account_id.clone(),
+                        timestamp: now,
+                        peers: my_peers.clone(),
+                    }.sign(vc.signer.as_ref()).unwrap())
+                }).collect();
+                // Insert node's own AccountData should never fail.
+                // We ignore the new data, because we trigger inserting
+                if let (_, Some(err)) = state.accounts_data.clone().insert(my_data).await {
```

Comment:
huh, this `.clone()` looks very confusing. perhaps

```diff
diff --git a/chain/network/src/accounts_data/mod.rs b/chain/network/src/accounts_data/mod.rs
index d2fbef173..d4af57aa9 100644
--- a/chain/network/src/accounts_data/mod.rs
+++ b/chain/network/src/accounts_data/mod.rs
@@ -215,7 +215,7 @@ impl Cache {
     /// Returns the data inserted and optionally a verification error.
     /// WriteLock is acquired only for the final update (after verification).
     pub async fn insert(
-        self: Arc<Self>,
+        self: &Arc<Self>,
         data: Vec<SignedAccountData>,
     ) -> (Vec<SignedAccountData>, Option<Error>) {
         let this = self.clone();

```

Reply:
ack, is there some convention around Arc<Self> receiver?

Reply:
Not really. Just that, if we are clonning it in the first line anyway, might as well &